### ISSUE TITLE
Hostname/IP does not match certificate's altnames: IP: 127.0.0.1 is not in the cert's list

### ISSUE THREAD

**Issue by rmuehlbauer** [bitnami/charts#10104]
### Name and Version

bitnami/mongodb:12.1.2

### What steps will reproduce the bug?

Upgraded mongodb:11.1.10 to 12.1.2 
Upgrade worked without problems but afterwards mongodb container keeps restarting due to failed liveness probes

### Are you using any custom parameters or values?

only part of values.yaml that seems to be interesting in this case

```
architecture: replicaset
## @param useStatefulSet Set to true to use a StatefulSet instead of a Deployment (only when `architecture=standalone`)
##
useStatefulSet: false
## MongoDB(&reg;) Authentication parameters
##
tls:
  ## @param tls.enabled Enable MongoDB(&reg;) TLS support between nodes in the cluster as well as between mongo clients and nodes
  ##
  enabled: true
  ## @param tls.autoGenerated Generate a custom CA and self-signed certificates
  ##
  autoGenerated: false
  ## @param tls.existingSecret Existing secret with TLS certificates (keys: `mongodb-ca-cert`, `mongodb-ca-key`, `client-pem`)
  ## NOTE: When it's set it will disable certificate creation
  ##
  existingSecret: ""
  ## Add Custom CA certificate
  ## @param tls.caCert Custom CA certificated (base64 encoded)
  ## @param tls.caKey CA certificate private key (base64 encoded)
  ##
  caCert: my-base64-encoded-CA-Cert
  caKey: my-base64-encoded-CA-Key
  ## Bitnami Nginx image
  ## @param tls.image.registry Init container TLS certs setup image registry
  ## @param tls.image.repository Init container TLS certs setup image repository
  ## @param tls.image.tag Init container TLS certs setup image tag (immutable tags are recommended)
  ## @param tls.image.pullPolicy Init container TLS certs setup image pull policy
  ## @param tls.image.pullSecrets Init container TLS certs specify docker-registry secret names as an array
  ## @param tls.extraDnsNames Add extra dns names to the CA, can solve x509 auth issue for pod clients
  ##
  image:
    registry: docker.io
    repository: bitnami/nginx
    tag: 1.21.6-debian-10-r103
    pullPolicy: IfNotPresent
    ## Optionally specify an array of imagePullSecrets.
    ## Secrets must be manually created in the namespace.
    ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
    ## e.g:
    ## pullSecrets:
    ##   - myRegistryKeySecretName
    ##
    pullSecrets: []

  ## e.g:
  ## extraDnsNames
  ##   "DNS.6": "$my_host"
  ##   "DNS.7": "$test"
  ##
  extraDnsNames: []
  
  ## @param tls.mode Allows to set the tls mode which should be used when tls is enabled (options: `allowTLS`, `preferTLS`, `requireTLS`)
  ##
  mode: requireTLS
## @param hostAliases Add deployment host aliases
## https://kubernetes.io/docs/concepts/services-networking/add-entries-to-pod-etc-hosts-with-host-aliases/
##
hostAliases: []
## @param replicaSetName Name of the replica set (only when `architecture=replicaset`)
## Ignored when mongodb.architecture=standalone
##
replicaSetName: replica-name
## @param replicaSetHostnames Enable DNS hostnames in the replicaset config (only when `architecture=replicaset`)
## Ignored when mongodb.architecture=standalone
## Ignored when externalAccess.enabled=true
##
replicaSetHostnames: true
## @param enableIPv6 Switch to enable/disable IPv6 on MongoDB(&reg;)
## ref: https://github.com/bitnami/bitnami-docker-mongodb/blob/master/README.md#enabling/disabling-ipv6
##
## configuration: |-

extraFlags: 
  - "--clusterAuthMode=x509"


externalAccess:
  ## @param externalAccess.enabled Enable Kubernetes external cluster access to MongoDB(&reg;) nodes (only for replicaset architecture)
  ##
  enabled: true
  ## External IPs auto-discovery configuration
  ## An init container is used to auto-detect LB IPs or node ports by querying the K8s API
  ## Note: RBAC might be required
  ##
  ## Parameters to configure K8s service(s) used to externally access MongoDB(&reg;)
  ## A new service per broker will be created
  ##
  service:
    ## @param externalAccess.service.type Kubernetes Service type for external access. Allowed values: NodePort, LoadBalancer or ClusterIP
    ##
    type: LoadBalancer
    ## @param externalAccess.service.portName MongoDB(&reg;) port name used for external access when service type is LoadBalancer
    ##
    portName: "mongodb"
    ## @param externalAccess.service.ports.mongodb MongoDB(&reg;) port used for external access when service type is LoadBalancer
    ##
    ports:
      mongodb: 27017
    ## @param externalAccess.service.loadBalancerIPs Array of load balancer IPs for MongoDB(&reg;) nodes
    ## Example:
    ## loadBalancerIPs:
    ##   - X.X.X.X
    ##   - Y.Y.Y.Y
    ##
    loadBalancerIPs:
    ## @param externalAccess.service.loadBalancerSourceRanges Address(es) that are allowed when service is LoadBalancer
      - internal-IP1
      - internal-IP2
      - internal-IP3
    ## ref: https://kubernetes.io/docs/tasks/access-application-cluster/configure-cloud-provider-firewall/#restrict-access-for-loadbalancer-service
    ## Example:
    ## loadBalancerSourceRanges:
    ## - 10.10.10.0/24
    ##
    loadBalancerSourceRanges: []
    ## @param externalAccess.service.externalTrafficPolicy MongoDB(&reg;) service external traffic policy
    ## ref https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip
    ##
    externalTrafficPolicy: Local
    ## @param externalAccess.service.nodePorts Array of node ports used to configure MongoDB(&reg;) advertised hostname when service type is NodePort
    ## Example:
    ## nodePorts:
    ##   - 30001
    ##   - 30002
    ##
    nodePorts: []
    ## @param externalAccess.service.domain Domain or external IP used to configure MongoDB(&reg;) advertised hostname when service type is NodePort
    ## If not specified, the container will try to get the kubernetes node external IP
    ## e.g:
    ## domain: mydomain.com
    ##
    domain: ""
    ## @param externalAccess.service.extraPorts Extra ports to expose (normally used with the `sidecar` value)
    ##
    extraPorts: []
    ## @param externalAccess.service.annotations Service annotations for external access
    ##
    annotations:
      networking.gke.io/load-balancer-type: "Internal"
    ## @param externalAccess.service.sessionAffinity Control where client requests go, to the same pod or round-robin
    ## Values: ClientIP or None
    ## ref: https://kubernetes.io/docs/user-guide/services/
    ##
    sessionAffinity: None
    ## @param externalAccess.service.sessionAffinityConfig Additional settings for the sessionAffinity
    ## sessionAffinityConfig:
    ##   clientIP:
    ##     timeoutSeconds: 300
    ##
    sessionAffinityConfig: {}
```

### What is the expected behavior?

mongodb container should not keep restarting after upgrade

### What do you see instead?

mongodb liveness probe fails

```
MongoServerSelectionError: Hostname/IP does not match certificate's altnames: IP: 127.0.0.1 is not in the cert's list: IP1, IP2, IP3, IP4
  Warning       Unhealthy  83s  kubelet  Liveness probe failed: Current Mongosh Log ID:  627a6c55f90720115af9fe1f
Connecting to:             mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+1.3.1
```

### Additional information

before upgrading i had
tls.autoGenerated: false

which i had to change to 
tls.autoGenerated: true

cause otherwise the "generate-tls-certs" init container does not start after upgrading, because the script /bitnami/scripts/generate-certs.sh does not exist in this case.

**Comment by rmuehlbauer** [comment]
after some more debugging and some research, it seems that mongosh is causing the problem.
The error message "Hostname/IP does not match certificate's altnames: IP: 127.0.0.1 is not in the cert's list" is caused by mongosh, which is used in the script "ping-mongodb.sh" for liveness probing.

The script executed for liveness checking
```
 #!/bin/bash
    TLS_OPTIONS='--tls --tlsCertificateKeyFile=/certs/mongodb.pem --tlsCAFile=/certs/mongodb-ca-cert'
    mongosh  $TLS_OPTIONS --port $MONGODB_PORT_NUMBER --eval "db.adminCommand('ping')"
```
tries to connect to a mongodb on 127.0.0.1
As 127.0.0.1 actually is not in the servers certificate list of ip's but only added as a alt name to the certificate, mongosh does not establish the connection.

I think the easiest way to fix this might be a little change to the liveness probe - basically only adding "--host localhost", which uses the dns name and not the ip address to connect to the local instance. Like so:
```
#!/bin/bash
    TLS_OPTIONS='--tls --tlsCertificateKeyFile=/certs/mongodb.pem --tlsCAFile=/certs/mongodb-ca-cert'
    mongosh  $TLS_OPTIONS --port $MONGODB_PORT_NUMBER --eval "db.adminCommand('ping')"
```

I was able to solve the problem this way.

**Comment by rmuehlbauer** [comment]
Hi @javsalgar, I think your pull request #9916 requires another little change - at least from my point of view .. please see above.
tldr; mongosh used in liveness probe refuses to connect to ip "127.0.0.1" but works well with "localhost" due to certificate checking failure. Imho, the easiest way to solve this might be to change ping-mongodb.sh in the common-scripts-cm.yaml - basically only add "--host localhost":
```
ping-mongodb.sh: |
    #!/bin/bash
    {{- if .Values.tls.enabled }}
    TLS_OPTIONS='--tls --tlsCertificateKeyFile=/certs/mongodb.pem --tlsCAFile=/certs/mongodb-ca-cert'
    {{- end }}
    mongosh  $TLS_OPTIONS --host localhost --port $MONGODB_PORT_NUMBER --eval "db.adminCommand('ping')"
  {{- if and .Values.tls.enabled .Values.tls.autoGenerated }}
```

**Comment by joancafom** [comment]
Hi @rmuehlbauer

Thanks for reporting! I am actively looking into this and will provide you with comments during the morning 😁

**Comment by joancafom** [comment]
Hi again,

After briefly checking this, this totally looks like an issue on the chart's side. I have been able to reproduce this using the following custom values (based on the ones you provided):

```yaml
architecture: replicaset
tls:
  enabled: true
```

This leads to an unstable state, making pods constantly crash and restart (as reported)

```
$ kubectl get pods
NAME                      READY   STATUS    RESTARTS   AGE
delta-mongodb-0           1/1     Running   2          3m34s
delta-mongodb-1           0/1     Running   1          2m59s
delta-mongodb-arbiter-0   1/1     Running   0          3m34s

$ kubectl describe pod/delta-mongodb-0
Connecting to:             mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+1.3.1
MongoServerSelectionError: Hostname/IP does not match certificate's altnames: IP: 127.0.0.1 is not in the cert's list: 10.99.24.3
```

> tldr; mongosh used in liveness probe refuses to connect to ip "127.0.0.1" but works well with "localhost" due to certificate checking failure. Imho, the easiest way to solve this might be to change ping-mongodb.sh in the common-scripts-cm.yaml - basically only add "--host localhost":

Though your proposed change works, I believe the issue is with either the certificates themselves or the way mongo checks their validity. From a theoretical point of view, it would not matter whether to use `127.0.0.1` or `localhost` if both entries appear in the SAN section.

After checking the certificate, it seems the error is related to how the `127.0.0.1` is added to the SAN section:

```
        X509v3 extensions:
            X509v3 Basic Constraints:
                CA:FALSE
            X509v3 Key Usage:
                Digital Signature, Non Repudiation, Key Encipherment
            X509v3 Subject Alternative Name:
                DNS:beta-mongodb-headless, DNS:beta-mongodb-0, DNS:beta-mongodb-0.beta-mongodb-headless.default.svc.cluster.local, DNS:localhost, DNS:127.0.0.1, IP Address:10.99.24.3
``` 

It should be added as an`IP Address` field instead of a `DNS` one. With that change, everything works like a charm:

```bash 
$ kubectl get pods
NAME                      READY   STATUS    RESTARTS   AGE
gamma-mongodb-0           1/1     Running   0          34m
gamma-mongodb-1           1/1     Running   0          33m
gamma-mongodb-arbiter-0   1/1     Running   0          34m
```

I will create a PR addressing this issue, thank you so much for reporting!

**Comment by rmuehlbauer** [comment]
I think something is still not right.

I've upgraded my chart to the latest version 12.1.5 and I can see the changed "generate-certs.sh" now has the extra alt name "IP.1 = 127.0.0.1" in it.
Anyhow, the resulting certificate does NOT have the 127.0.0.1 in it:
```
X509v3 extensions:
            X509v3 Basic Constraints:
                CA:FALSE
            X509v3 Key Usage:
                Digital Signature, Non Repudiation, Key Encipherment
            X509v3 Subject Alternative Name:
                DNS:mongodb-b2b-dev-headless, DNS:mongodb-b2b-dev-2, DNS:mongodb-b2b-dev-2.mongodb-b2b-dev-headless.b2b.svc.cluster.local, DNS:localhost, IP Address:10.43.0.34, IP Address:10.43.0.205, IP Address:10.43.0.206, IP Address:10.43.0.207
```

As you can see, there are 4 ips in the certificate - the first one is the pods ip and the 3 additional ones come from my values.yaml from the externalAccess.service.loadBalancerIPs section...and I think this causes the issue here and overwrites the "IP.1 = 127.0.0.1".

In the "generate-certs.sh" we can find the following lines:
```
for ip in "${additional_ips[@]}"; do
        cat >>/certs/openssl.cnf <<EOL
    IP.$index = $ip
```

So, if there are any additional loadBalancerIPs, they are going to overwrite "IP.1 = 127.0.0.1" with their value and the liveness check will fail again with the error message
```
Hostname/IP does not match certificate's altnames: IP: 127.0.0.1 is not in the cert's list: 10.43.0.34, 10.43.0.205, 10.43.0.206, 10.43.0.207
```

I'm not sure, where to go from here....one possible approach would be, to respect the ips for the loadBalancers and add 127.0.0.1 as an additional one.
Another approach might still be, to add "--host localhost" to the mongosh ping script - like I've already suggested and also can confirm working in my setup (at least).

**Comment by joancafom** [comment]
Hi @rmuehlbauer 

I see that you already opened a new issue, let's continue the discussion there

Thanks

**Comment by wjentner** [comment]
We are also having this issue with chart version 11.2.0. 
I am hesitant to upgrade to v12 since it will upgrade mongodb from v4 to v5 -- is the new chart still compatible with mongodb v4?
